// @langchain/langgraph-checkpoint-postgres/dist/store/modules/database-core.cjs

//#region src/store/modules/database-core.ts
/**
* Core database operations and connection management.
* Shared by all modules to avoid duplication.
*/
var DatabaseCore = class {
	pool;
	schema;
	ttlConfig;
	indexConfig;
	textSearchLanguage;
	constructor(pool, schema, ttlConfig, indexConfig, textSearchLanguage) {
		this.pool = pool;
		this.schema = schema;
		this.ttlConfig = ttlConfig;
		this.indexConfig = indexConfig;
		this.textSearchLanguage = textSearchLanguage || "english";
	}
	async withClient(operation) {
		const client = await this.pool.connect();
		try {
			return await operation(client);
		} finally {
			client.release();
		}
	}
	calculateExpiresAt(ttl) {
		const effectiveTtl = ttl ?? this.ttlConfig?.defaultTtl;
		if (!effectiveTtl) return null;
		return new Date(Date.now() + effectiveTtl * 60 * 1e3);
	}
	async refreshTtl(client, namespacePath, key) {
		if (!this.ttlConfig?.refreshOnRead) return;
		const expiresAt = this.calculateExpiresAt();
		if (expiresAt) await client.query(`
        UPDATE "${this.schema}".store 
        SET expires_at = $3, updated_at = CURRENT_TIMESTAMP
        WHERE namespace_path = $1 AND key = $2
      `, [
			namespacePath,
			key,
			expiresAt
		]);
	}
};

//#endregion
exports.DatabaseCore = DatabaseCore;
