// @langchain/langgraph-checkpoint-postgres/dist/store/modules/search-operations.js
import { validateNamespace } from "./utils.js";
import { QueryBuilder } from "./query-builder.js";

//#region src/store/modules/search-operations.ts
/**
* Handles all search operations: basic search, vector search, hybrid search.
*/
var SearchOperations = class {
	constructor(core, vectorOps) {
		this.core = core;
		this.vectorOps = vectorOps;
	}
	async executeSearch(client, operation) {
		validateNamespace(operation.namespacePrefix);
		const namespacePath = operation.namespacePrefix.join(":");
		const { filter, limit = 10, offset = 0, query } = operation;
		if (query && this.core.indexConfig) return this.executeVectorSearch(client, operation);
		if (query && !this.core.indexConfig) return (await this.textSearch(operation.namespacePrefix, {
			query,
			filter,
			limit,
			offset
		})).map((item) => ({
			namespace: item.namespace,
			key: item.key,
			value: item.value,
			createdAt: item.createdAt,
			updatedAt: item.updatedAt
		}));
		let sqlQuery = `
      SELECT namespace_path, key, value, created_at, updated_at
      FROM "${this.core.schema}".store
      WHERE namespace_path LIKE $1
        AND (expires_at IS NULL OR expires_at > CURRENT_TIMESTAMP)
    `;
		const params = [`${namespacePath}%`];
		let paramIndex = 2;
		if (filter && Object.keys(filter).length > 0) {
			const { conditions, newParamIndex } = QueryBuilder.buildFilterConditions(filter, params, paramIndex);
			if (conditions.length > 0) {
				sqlQuery += ` AND (${conditions.join(" AND ")})`;
				paramIndex = newParamIndex;
			}
		}
		sqlQuery += ` ORDER BY created_at DESC LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`;
		params.push(limit, offset);
		return (await client.query(sqlQuery, params)).rows.map((row) => ({
			namespace: row.namespace_path.split(":"),
			key: row.key,
			value: row.value,
			createdAt: row.created_at,
			updatedAt: row.updated_at
		}));
	}
	async executeVectorSearch(client, operation) {
		if (!this.core.indexConfig || !operation.query) return [];
		validateNamespace(operation.namespacePrefix);
		const namespacePath = operation.namespacePrefix.join(":");
		const { filter, limit = 10, offset = 0, query } = operation;
		const queryEmbedding = await this.vectorOps.generateQueryEmbedding(query);
		if (queryEmbedding.length !== this.core.indexConfig.dims) throw new Error(`Query embedding dimension mismatch: expected ${this.core.indexConfig.dims}, got ${queryEmbedding.length}`);
		let sqlQuery = `
      SELECT DISTINCT 
        s.namespace_path, 
        s.key, 
        s.value, 
        s.created_at, 
        s.updated_at,
        MIN(v.embedding <=> $2) as similarity_score
      FROM "${this.core.schema}".store s
      JOIN "${this.core.schema}".store_vectors v ON s.namespace_path = v.namespace_path AND s.key = v.key
      WHERE s.namespace_path LIKE $1
        AND (s.expires_at IS NULL OR s.expires_at > CURRENT_TIMESTAMP)
    `;
		const params = [`${namespacePath}%`, `[${queryEmbedding.join(",")}]`];
		let paramIndex = 3;
		if (filter && Object.keys(filter).length > 0) {
			const { conditions, newParamIndex } = QueryBuilder.buildFilterConditions(filter, params, paramIndex);
			if (conditions.length > 0) {
				const adjustedConditions = conditions.map((condition) => condition.replace(/value ->/g, "s.value ->"));
				sqlQuery += ` AND (${adjustedConditions.join(" AND ")})`;
				paramIndex = newParamIndex;
			}
		}
		sqlQuery += ` 
      GROUP BY s.namespace_path, s.key, s.value, s.created_at, s.updated_at
      ORDER BY similarity_score ASC 
      LIMIT $${paramIndex} OFFSET $${paramIndex + 1}
    `;
		params.push(limit, offset);
		return (await client.query(sqlQuery, params)).rows.map((row) => ({
			namespace: row.namespace_path.split(":"),
			key: row.key,
			value: row.value,
			createdAt: row.created_at,
			updatedAt: row.updated_at,
			score: 1 - parseFloat(row.similarity_score)
		}));
	}
	async textSearch(namespacePrefix, options = {}) {
		return this.core.withClient(async (client) => {
			validateNamespace(namespacePrefix);
			const namespacePath = namespacePrefix.join(":");
			const { filter, limit = 10, offset = 0, query, refreshTtl } = options;
			let sqlQuery = `
        SELECT 
          namespace_path, 
          key, 
          value, 
          created_at, 
          updated_at,
          CASE 
            WHEN $2::text IS NOT NULL THEN 
              ts_rank(to_tsvector($3::regconfig, value::text), plainto_tsquery($3::regconfig, $2::text))
            ELSE 0
          END as score
        FROM "${this.core.schema}".store
        WHERE namespace_path LIKE $1
          AND (expires_at IS NULL OR expires_at > CURRENT_TIMESTAMP)
      `;
			const params = [
				`${namespacePath}%`,
				query || null,
				this.core.textSearchLanguage
			];
			let paramIndex = 4;
			if (filter && Object.keys(filter).length > 0) {
				const { conditions, newParamIndex } = QueryBuilder.buildFilterConditions(filter, params, paramIndex);
				if (conditions.length > 0) {
					sqlQuery += ` AND (${conditions.join(" AND ")})`;
					paramIndex = newParamIndex;
				}
			}
			if (query) {
				sqlQuery += ` AND (
          to_tsvector($3::regconfig, value::text) @@ plainto_tsquery($3::regconfig, $2::text)
          OR value::text ILIKE $${paramIndex}
        )`;
				params.push(`%${query}%`);
				paramIndex += 1;
			}
			if (query) sqlQuery += ` ORDER BY score DESC, updated_at DESC`;
			else sqlQuery += ` ORDER BY updated_at DESC`;
			sqlQuery += ` LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`;
			params.push(limit, offset);
			const items = (await client.query(sqlQuery, params)).rows.map((row) => ({
				namespace: row.namespace_path.split(":"),
				key: row.key,
				value: row.value,
				createdAt: row.created_at || /* @__PURE__ */ new Date(),
				updatedAt: row.updated_at || /* @__PURE__ */ new Date(),
				score: row.score || void 0
			}));
			if (refreshTtl || this.core.ttlConfig?.refreshOnRead) for (const item of items) await this.core.refreshTtl(client, item.namespace.join(":"), item.key);
			return items;
		});
	}
	async vectorSearch(namespacePrefix, query, options = {}) {
		if (!this.core.indexConfig) throw new Error("Vector search not configured. Please provide an IndexConfig when creating the store.");
		return this.core.withClient(async (client) => {
			validateNamespace(namespacePrefix);
			const namespacePath = namespacePrefix.join(":");
			const { filter, limit = 10, offset = 0, similarityThreshold = 0, distanceMetric = "cosine" } = options;
			const queryEmbedding = await this.vectorOps.generateQueryEmbedding(query);
			if (queryEmbedding.length !== this.core.indexConfig.dims) throw new Error(`Query embedding dimension mismatch: expected ${this.core.indexConfig.dims}, got ${queryEmbedding.length}`);
			let distanceOp;
			let scoreTransform;
			switch (distanceMetric) {
				case "l2":
					distanceOp = "<->";
					scoreTransform = "1 / (1 + MIN(v.embedding <-> $2))";
					break;
				case "inner_product":
					distanceOp = "<#>";
					scoreTransform = "MIN(v.embedding <#> $2)";
					break;
				default:
					distanceOp = "<=>";
					scoreTransform = "1 - MIN(v.embedding <=> $2)";
					break;
			}
			let sqlQuery = `
        SELECT DISTINCT 
          s.namespace_path, 
          s.key, 
          s.value, 
          s.created_at, 
          s.updated_at,
          ${scoreTransform} as similarity_score
        FROM "${this.core.schema}".store s
        JOIN "${this.core.schema}".store_vectors v ON s.namespace_path = v.namespace_path AND s.key = v.key
        WHERE s.namespace_path LIKE $1
          AND (s.expires_at IS NULL OR s.expires_at > CURRENT_TIMESTAMP)
      `;
			const params = [`${namespacePath}%`, `[${queryEmbedding.join(",")}]`];
			let paramIndex = 3;
			if (similarityThreshold > 0) {
				if (distanceMetric === "inner_product") sqlQuery += ` AND v.embedding <#> $2 >= $${paramIndex}`;
				else sqlQuery += ` AND v.embedding ${distanceOp} $2 <= $${paramIndex}`;
				params.push(distanceMetric === "cosine" ? 1 - similarityThreshold : similarityThreshold);
				paramIndex += 1;
			}
			if (filter && Object.keys(filter).length > 0) {
				const { conditions, newParamIndex } = QueryBuilder.buildFilterConditions(filter, params, paramIndex);
				if (conditions.length > 0) {
					const adjustedConditions = conditions.map((condition) => condition.replace(/value ->/g, "s.value ->"));
					sqlQuery += ` AND (${adjustedConditions.join(" AND ")})`;
					paramIndex = newParamIndex;
				}
			}
			sqlQuery += ` 
        GROUP BY s.namespace_path, s.key, s.value, s.created_at, s.updated_at
        ORDER BY similarity_score DESC
        LIMIT $${paramIndex} OFFSET $${paramIndex + 1}
      `;
			params.push(limit, offset);
			return (await client.query(sqlQuery, params)).rows.map((row) => ({
				namespace: row.namespace_path.split(":"),
				key: row.key,
				value: row.value,
				createdAt: row.created_at || /* @__PURE__ */ new Date(),
				updatedAt: row.updated_at || /* @__PURE__ */ new Date(),
				score: parseFloat(row.similarity_score)
			}));
		});
	}
	async hybridSearch(namespacePrefix, query, options = {}) {
		if (!this.core.indexConfig) throw new Error("Vector search not configured. Please provide an IndexConfig when creating the store.");
		return this.core.withClient(async (client) => {
			validateNamespace(namespacePrefix);
			const namespacePath = namespacePrefix.join(":");
			const { filter, limit = 10, offset = 0, vectorWeight = .7, similarityThreshold = 0 } = options;
			const queryEmbedding = await this.vectorOps.generateQueryEmbedding(query);
			if (queryEmbedding.length !== this.core.indexConfig.dims) throw new Error(`Query embedding dimension mismatch: expected ${this.core.indexConfig.dims}, got ${queryEmbedding.length}`);
			let sqlQuery = `
        SELECT DISTINCT 
          s.namespace_path, 
          s.key, 
          s.value, 
          s.created_at, 
          s.updated_at,
          (
            $3 * (1 - MIN(v.embedding <=> $2)) + 
            (1 - $3) * ts_rank(to_tsvector($6::regconfig, s.value::text), plainto_tsquery($6::regconfig, $4))
          ) as hybrid_score
        FROM "${this.core.schema}".store s
        JOIN "${this.core.schema}".store_vectors v ON s.namespace_path = v.namespace_path AND s.key = v.key
        WHERE s.namespace_path LIKE $1
          AND (s.expires_at IS NULL OR s.expires_at > CURRENT_TIMESTAMP)
          AND (
            to_tsvector($6::regconfig, s.value::text) @@ plainto_tsquery($6::regconfig, $4)
            OR v.embedding <=> $2 <= $5
          )
      `;
			const params = [
				`${namespacePath}%`,
				`[${queryEmbedding.join(",")}]`,
				vectorWeight,
				query,
				1 - similarityThreshold,
				this.core.textSearchLanguage
			];
			let paramIndex = 7;
			if (filter && Object.keys(filter).length > 0) {
				const { conditions, newParamIndex } = QueryBuilder.buildFilterConditions(filter, params, paramIndex);
				if (conditions.length > 0) {
					const adjustedConditions = conditions.map((condition) => condition.replace(/value ->/g, "s.value ->"));
					sqlQuery += ` AND (${adjustedConditions.join(" AND ")})`;
					paramIndex = newParamIndex;
				}
			}
			sqlQuery += ` 
        GROUP BY s.namespace_path, s.key, s.value, s.created_at, s.updated_at
        ORDER BY hybrid_score DESC
        LIMIT $${paramIndex} OFFSET $${paramIndex + 1}
      `;
			params.push(limit, offset);
			return (await client.query(sqlQuery, params)).rows.map((row) => ({
				namespace: row.namespace_path.split(":"),
				key: row.key,
				value: row.value,
				createdAt: row.created_at || /* @__PURE__ */ new Date(),
				updatedAt: row.updated_at || /* @__PURE__ */ new Date(),
				score: parseFloat(row.hybrid_score)
			}));
		});
	}
};

//#endregion
export { SearchOperations };
